nomad/plan_apply_test.go

```diff
@@ -228,6 +230,154 @@ func TestPlanApply_applyPlan(t *testing.T) {
 	assert.Equal(index, evalOut.ModifyIndex)
 }
 
+func TestPlanApply_applyPlanWithNormalizedAllocs(t *testing.T) {
+	t.Parallel()
+	s1 := TestServer(t, func(c *Config) {
+		c.Build = "0.9.1"
+	})
+	defer s1.Shutdown()
+	testutil.WaitForLeader(t, s1.RPC)
+
+	// Register node
+	node := mock.Node()
+	testRegisterNode(t, s1, node)
+
+	// Register a fake deployment
+	oldDeployment := mock.Deployment()
+	if err := s1.State().UpsertDeployment(900, oldDeployment); err != nil {
+		t.Fatalf("UpsertDeployment failed: %v", err)
+	}
+
+	// Create a deployment
+	dnew := mock.Deployment()
+
+	// Create a deployment update for the old deployment id
+	desiredStatus, desiredStatusDescription := "foo", "bar"
+	updates := []*structs.DeploymentStatusUpdate{
+		{
+			DeploymentID:      oldDeployment.ID,
+			Status:            desiredStatus,
+			StatusDescription: desiredStatusDescription,
+		},
+	}
+
+	// Register allocs, deployment and deployment update
+	alloc := mock.Alloc()
+	stoppedAlloc := mock.Alloc()
+	stoppedAllocDiff := &structs.Allocation{
+		ID:                 stoppedAlloc.ID,
+		DesiredDescription: "Desired Description",
+		ClientStatus:       structs.AllocClientStatusLost,
+	}
+	preemptedAlloc := mock.Alloc()
+	preemptedAllocDiff := &structs.Allocation{
+		ID:                    preemptedAlloc.ID,
+		PreemptedByAllocation: alloc.ID,
+	}
+	s1.State().UpsertJobSummary(1000, mock.JobSummary(alloc.JobID))
+	s1.State().UpsertAllocs(1100, []*structs.Allocation{stoppedAlloc, preemptedAlloc})
+	// Create an eval
+	eval := mock.Eval()
+	eval.JobID = alloc.JobID
+	if err := s1.State().UpsertEvals(1, []*structs.Evaluation{eval}); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	timestampBeforeCommit := time.Now().UTC().UnixNano()
+	planRes := &structs.PlanResult{
+		NodeAllocation: map[string][]*structs.Allocation{
+			node.ID: {alloc},
+		},
+		NodeUpdate: map[string][]*structs.Allocation{
+			stoppedAlloc.NodeID: {stoppedAllocDiff},
+		},
+		NodePreemptions: map[string][]*structs.Allocation{
+			preemptedAlloc.NodeID: {preemptedAllocDiff},
+		},
+		Deployment:        dnew,
+		DeploymentUpdates: updates,
+	}
+
+	// Snapshot the state
+	snap, err := s1.State().Snapshot()
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// Create the plan with a deployment
+	plan := &structs.Plan{
+		Job:               alloc.Job,
+		Deployment:        dnew,
+		DeploymentUpdates: updates,
+		EvalID:            eval.ID,
+	}
+
+	// Apply the plan
+	future, err := s1.applyPlan(plan, planRes, snap)
+	assert := assert.New(t)
```

Comment:
This is mostly following the same pattern as the older denormalized case did (and the same is true for other tests which have a denormalized version from the past). Having said that, I think it's probably not a good idea for only the first failure to matter in a test. The reason being that the programmer doesn't want to go through a cycle wherein they have to go back and forth while fixing a test to uncover all the issues. Especially for a test that tests multiple somewhat independent objects as this one does. In cases where only one object is tested, and that one was nil and there's no point testing it for attributes, it's better in cases like those to stop running the test.

Reply:
@arsh - For historic context, we used assert for a while in the past (which is why some old tests do this) and more recently have pretty much standardized on require. In this test for instance, if the plan apply failed and returned an error there's no point in examining allocations expected to be created from that plan. If an alloc look up from the state store errored, the immediate next line checking for some state in the alloc will also fail. More often than not require is what you want anyway, so we've tried to standardize on that for new code.

Reply:
My thought process was that plan apply does multiple inserts, and it could fail at any point, so its probably a good idea to know what failed and what didn't. And, doing a nil check for require makes perfect sense, but it's hard to do a require and flip to assert (little value, maintain order of objects asserted). I'll change the test to use require, since that is what the team has standardized on.